### ISSUE TITLE
Feature: Add a way to mock a constructor with defined values

### ISSUE THREAD

**Issue by Tryle** [mockk/mockk#209]
### Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Expected Behavior

Actually we can [mock constructors](https://github.com/mockk/mockk#constructor-mocks) but there is no way to handle constructors parameters. Something like:

```kotlin
class MockCls(val minValue: Int) {
  fun add(a: Int, b: Int) = minValue + a + b
}

mockkConstructor(MockCls::class)

every { anyConstructed<MockCls>(10).add(1, 2) } returns 14

assertEquals(14, MockCls(10).add(1, 2)) // note new object is created

verify { anyConstructed<MockCls>(10).add(1, 2) } // success
verify { anyConstructed<MockCls>(5).add(1, 2) } // fail
```

For reference there is [a question on SO](https://stackoverflow.com/questions/53902932/can-we-mock-constructor-and-check-its-parameters) about that.

**Comment by mpost** [comment]
I am facing the same requirement to retrieve the constructor arguments for the created class. It is good to see a feature request which is not closed immediately so it seems to be at least feasible. Hope to see it soon.

**Comment by erikhuizinga** [comment]
I'm not sure, but if I read the readme correctly, the current implementation of `mockkConstructor` can only create one mock for all constructed instances, am I right? By that one mock, I mean the object you get through the `anyConstructed` function.

This, of course, is not ideal for true power to the user. Ideally we could do something like this:


```kotlin
// Note this class has two constructors
class Cls() {
    constructor(a: Any?) : this()
}

@Test 
fun `some test`() = mockkConstructors(Cls::class) {
    val mock1 = mockk<Cls>()
    val mock2 = mockk<Cls>()
	val a = 0

	every { Cls() } returns mock1
	every { Cls(a) } returns mock2
}
```

Notes:
 - Note the 's' in `mockkConstructors`: this would mock all available constructors, which can potentially be many.
   - OTOH, as a MockK user I'd want the ability to specifically mock a constructor with a specific signature.
 - Note that I return different instances in different cases (in this case mockks, but of course, these can be any valid instance).

**Comment by oleksiyp** [comment]
Yes, you are right. The underlying mocking stub is only one. And all the calls from created objects via `mockkConstructor` are delegated to this stub.

Returning arbitrary object/mocks as a result of the constructor calls is not possible. But attaching different stubs is possible.

I've made an attempt to implement a variety of constructors in a simple way: https://github.com/mockk/mockk/pull/267
But seems full rewrit\e is needed. This implementation lacks a few things. 

I have an idea that `constructor call` should be somehow indistinguishable from `method` call at the level of DSL engine. This will unify it the way you can apply all the machinery with `matchers`.

**Comment by erikhuizinga** [comment]
@stale[bot] come on ~~man~~ bot! No activity means no activity, but not that an issue should be closed! This way active issues or even bugs might lose visibility.

**Comment by darrengyl** [comment]
We need this! It will be really useful to test constructors with parameters

**Comment by bboyairwreck** [comment]
can i asked why this was closed? This would be really nice functionality that has a real use cases. Stop closing this ticket bot

**Comment by oleksiyp** [comment]
Check #356

**Comment by espertus** [comment]
I would like this too.

**Comment by Richrocksmy** [comment]
In relation to this, when mocking constructors that are not the default constructor ie they have parameters the mocking doesn't appear to work and instead the real code gets called.

I think that adding the ability to mock constructors individually would fix this issue.

**Comment by marklapasa** [comment]
bump - would like to see this too

**Comment by rkokhatskyi** [comment]
+1 for this feature

**Comment by thesultanster** [comment]
+1 for this as well

**Comment by davidcre** [comment]
+1 the current implementation of constructor mocking is not enough.
Thank you for considerign this feature 👍

**Comment by noah-iam** [comment]
I am facing the same issue while mocking the class with parameter constructor. 

`io.mockk.MockKException: No matching constructors found:
`

**Comment by miva2** [comment]
I tried to write this but apparently it's not possible
        every { SomeClass(param1, param2) } returns SomeClassMock

I haven't used MockK a lot yet but expected it to work like that.

**Comment by JureSencar** [comment]
+1

Already returning a pre-initialied (mocked) instance would be useful:
`every { SomeClass(value1) } returns someInstance`
`every { SomeClass(not(value1)) } returns someOtherInstance`

**Comment by anticafe** [comment]
+1 for this feature.

**Comment by distante** [comment]
I am also very interested in this.

**Comment by othiagosouto** [comment]
+1

**Comment by JureSencar** [comment]
+1

The workaround I use (in case I own the code) is to define a factory method with `@VisibleForTesting` to the companion object of the class and mockk that. Of course, changing source for the sake of testing is not nice and it is not an option if you don't own the code. Thus, I am definitively interested in this feature.

**Comment by kenaiX** [comment]
+1

**Comment by MartinDPutraTvlk** [comment]
+1

**Comment by JanCizmar** [comment]
+1

**Comment by om32** [comment]
+1

**Comment by EclipseOnFire** [comment]
+1

**Comment by campino-rocks** [comment]
+1

**Comment by Diones** [comment]
+1

**Comment by gregorgrisza** [comment]
+1

**Comment by konstantin-tkachuk** [comment]
+1

**Comment by felipezad** [comment]
+1

**Comment by FGuyet** [comment]
+1!

**Comment by olgachrostowska-meelogic** [comment]
+1!

**Comment by AshokKumarMandava** [comment]
+
1

**Comment by KevinBassaDevelopment** [comment]
+1

**Comment by Bebooo43** [comment]
+1

**Comment by erikhuizinga** [comment]
Has this partially/entirely been fixed in v1.11.0?

**Comment by Raibaz** [comment]
Yep, v1.11.0, just released, adds `constructedWith` that fixes this.
